mlem/cli/utils.py

```diff
@@ -0,0 +1,591 @@
+            default=default,
+            is_list=is_list,
+            is_mapping=is_mapping,
+            mapping_key_type=str,
+        )
+        return
+    if isinstance(type_, type) and issubclass(type_, BaseModel):
+        # BaseModel (including MlemABC non-root classes): reqursively get nested
+        yield from iterate_type_fields(type_, path, not required)
+        return
+    # probably primitive field
+    yield CliTypeField(
+        required=required,
+        allow_none=allow_none,
+        path=path,
+        type_=type_,
+        default=default,
+        help=help_,
+        is_list=is_list,
+        is_mapping=is_mapping,
+        mapping_key_type=str,
+    )
+
+
+def iterate_type_fields(
+    cls: Type[BaseModel], path: str = "", force_not_req: bool = False
+) -> Iterator[CliTypeField]:
+    """Recursively get CliTypeFields from BaseModel"""
+    field: ModelField
+    for name, field in sorted(
+        cls.__fields__.items(), key=lambda x: not x[1].required
+    ):
+        name = field.alias or name
+        if issubclass(cls, MlemObject) and name in MlemObject.__fields__:
+            # Skip base MlemObject fields
+            continue
+        if (
+            issubclass(cls, MlemABC)
+            and name in cls.__config__.exclude
+            or field.field_info.exclude
+        ):
+            # Skip excluded fields
+            continue
+        if name == "__root__":
+            fullname = path
+        else:
+            fullname = name if not path else f"{path}.{name}"
+
+        field_type = field.type_
+        # field.type_ is element type for collections/mappings
+
+        if not isinstance(field_type, type):
+            # Handle generics. Probably will break in complex cases
+            origin = get_origin(field_type)
+            if is_union(origin):
+                # get first type for union
+                generic_args = get_args(field_type)
+                field_type = generic_args[0]
+            if origin is list or origin is dict:
+                # replace with dynamic __root__: Dict/List model
+                field_type = create_model(
+                    display_as_type(field_type), __root__=(field_type, ...)
+                )
+            if field_type is Any:
+                field_type = anything(field_type)
+
+        if not isinstance(field_type, type):
+            # skip too complicated stuff
+            continue
+
+        yield from parse_type_field(
+            path=fullname,
+            type_=field_type,
+            help_=get_field_help(cls, name),
+            is_list=field.shape in LIST_LIKE_SHAPES,
+            is_mapping=field.shape in MAPPING_LIKE_SHAPES,
+            required=not force_not_req and bool(field.required),
+            allow_none=field.allow_none,
+            default=field.default,
+        )
+
```

Comment:
should we remove extra space in `Usage: ...`?
<img width="626" alt="Screen Shot 2022-08-30 at 12 20 51" src="https://user-images.githubusercontent.com/6797716/187355797-4cfbfeae-e9f9-444c-b79a-be6cf97d7dec.png">

Reply:
this one is a typer error. usage is actually `mlem server [options] server`, but typer uses rich console and it treats this string as a markup, so `[options]` is a link with empty name probably. we should report it, but I just cant stand typer's issue reporting template

Reply:
Ok, let's report it then :) I guess you can skip their template if you wish to)